pyannote/core/interval_tree.py

```diff
@@ -25,286 +25,84 @@
-                self.num += right.num
-                self.extent = (min(self.extent[0], right.extent[0]),
-                               max(self.extent[1], right.extent[1]))
-
-            self.extent = Segment(*self.extent)
-
-        def __repr__(self):
-            return '{extent: %s, num: %d}' % (repr(self.extent), self.num)
-
-    def length(self):
-        if self.root:
-            return self.root.metadata.num
-        else:
-            return 0
-
-    def kth(self, k):
-        """Get kth segment"""
-        if k < 0:
-            return _kth(self.root, self.root.num + k)
-        else:
-            return _kth(self.root, k)
-
-    def index(self, segment):
-        return _index(self.root, segment, 0)
+import collections as co
+import sortedcontainers as sc
+
+class SortedDict(sc.SortedDict):
+    length = dict.__len__
+
+class SortedSet(co.MutableSet):
+    "Sorted set intended only for segments."
+    def __init__(self, segments=()):
+        "Initialize sorted set from iterable of segments."
+        set_segments = set(segments)
+        self._set = set_segments
+        self._list = sc.SortedList(set_segments)
+        marks = (mark for segment in set_segments for mark in segment)
+        self._marks = sc.SortedList(marks)
+
+    def __contains__(self, segment):
+        return segment in self._set
+
+    def __iter__(self):
+        return iter(self._list)
+
+    def __len__(self):
+        return len(self._set)
+
+    length = __len__
+
+    def add(self, segment):
+        _set = self._set
+        if segment in _set:
+            return
+        _set.add(segment)
+        self._list.add(segment)
+        _marks = self._marks
+        _marks.add(segment.start)
+        _marks.add(segment.end)
+
+    def discard(self, segment):
+        _set = self._set
+        if segment not in _set:
+            return
+        _set.remove(segment)
+        self._list.remove(segment)
+        _marks = self._marks
+        _marks.remove(segment.start)
+        _marks.remove(segment.end)
 
     def extent(self):
-        if self.root:
-            return self.root.metadata.extent
+        if self._set:
+            _marks = self._marks
+            start = _marks[0]
+            end = _marks[-1]
+            return Segment(start=start, end=end)
         else:
             return Segment(start=np.inf, end=-np.inf)
 
-    def debug(self):
-        _debug(self.root, 0)
-
-    def overlapping(self, t):
-        """Get list of segments overlapping time t"""
-        return _overlapping(self.root, t)
-
-    def intersecting(self, segment):
-        return _intersecting(self.root, segment)
+    def kth(self, index):
+        return self._list[index]
 
     def co_iter(self, other):
-        for segment, other_segment in _co_iter(self.root, other.root):
-            yield segment, other_segment
+        for alpha in self._list:
+            temp = Segment(start=alpha.end, end=alpha.end)
+            iterable = other._list.irange(maximum=temp)
```

Comment:
Yes, but only if the list has at least X elements, and the difference in length between the two iterators is at least Y elements. I don't know X and Y exactly but I would guess X is tens of thousands of elements and Y is dozens or hundreds of elements. We would need a good set of benchmarks to find the tradeoff and without those this seemed simplest.

Iterating sorted lists is very fast and the "irange" method will maintain most of that speed. Internally that's because everything is already laid out in lists so it's about as fast as iterating lists. But getting the index of an element is much slower because a separate index table must then be maintained. The index table is well optimized but much slower than simple iterating. So there's a tradeoff. For small lists and small differences in iterable length, the index table is not worth maintaining. But for larger lists, reducing the number of intersects method calls will be much more valuable than maintaining the index table.

Maintaining a second list will also consume more memory so if lot's of SortedSet's are created then there may be other factors of concern. There's probably a small enough SortedSet size where you should use a native Python list, then a larger size where you promote that to a SortedList, and then an even larger size where you use two SortedLists.

The best option is probably to construct benchmarks where we can test these hypotheses. I may be wrong about my magnitudes for X and Y. Without those, the simple solution is probably "good enough".